File: ruby/text-mining/extract_category_map.rb
# -*- coding: utf-8 -*-

require 'json'

class Analyzer
  def initialize(args)
    @filename = args.shift || "category_map.txt"
    @exclude  = args.shift || "wordcount_exclude.txt"
    @category_num = 0
    read_from_exclude
  end

  def extract_map(category)
    hits = {}
    exclude_count = 0
    open(@filename) do |file|
      file.each_line do |line|
        word, counts, social, politics, international, economics, electro, sports, entertainment, science, standard_deviation = line.force_encoding("utf-8").strip.split("\t")
        array = [social.to_i, politics.to_i, international.to_i, economics.to_i, electro.to_i, sports.to_i, entertainment.to_i, science.to_i]
        unless array[@category_num].to_i == 0
          #if array.max < 100
          #if counts.to_i == array.max or standard_deviation.to_f < 0.4
          if standard_deviation.to_f < 10.0
            unless @exclude_words.include?(word)
              if word =~ /[一-龠]/
                hits.has_key?(word) ? hits[word] += array[@category_num].to_i * 3 : hits[word] = array[@category_num].to_i * 3
              elsif word =~ /^[A-Za-z].*/
                hits.has_key?(word) ? hits[word] += array[@category_num].to_i : hits[word] = array[@category_num].to_i
              end
            end
          end
        end
      end
    end
    @category_num += 1
    output(@category_num, category, hits)
    return hits
  end

  def start
    extract_map('category.social')
    extract_map('category.politics')
    extract_map('category.international')
    extract_map('category.economics')
    extract_map('category.electro')
    extract_map('category.sports')
    extract_map('category.entertainment')
    extract_map('category.science')
  end

  private

  def read_from_exclude
    @exclude_words = Array.new
    open(@exclude) do |file|
      file.each_line do |line|
        @exclude_words << line.force_encoding("utf-8").chomp
      end
    end
  end

  def output(key, tag, value)
    puts "#{key}\t#{tag}\t#{JSON.generate(value, opts = {:space => ' ', :indent => ' '})}"
  end
end

if __FILE__ == $0
  analyzer = Analyzer.new(ARGV)
  analyzer.start
end

File: python/text-mining/extract_category_map.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys, os
import json
import re
from collections import OrderedDict

class Analyzer:
    def __init__(self, args):
        self.filename = args[1]
        self.exclude  = args[2]
        self.exclude_list = []
        self.category_num = 0
        self.__read_from_exclude()

    def start(self):
        self.__extract_map('category.social')
        self.__extract_map('category.politics')
        self.__extract_map('category.international')
        self.__extract_map('category.economics')
        self.__extract_map('category.electro')
        self.__extract_map('category.sports')
        self.__extract_map('category.entertainment')
        self.__extract_map('category.science')

    def __output__(self, key, tag, value):
        print(key, tag, json.dumps(value, ensure_ascii=False), sep="\t")

    def __extract_map(self, category):
        self.dic = OrderedDict()
        file = open(self.filename, 'r')
        for line in file:
            word, counts, social, politics, international, economics, electro, sports, entertainment, science, standard_deviation = line.rstrip().split("\t")
            array = [int(social), int(politics), int(international), int(economics), int(electro), int(sports), int(entertainment), int(science)]
            if not array[self.category_num] == 0:
                if float(standard_deviation) < 10.0:
                    if not word in self.exclude_list:
                        r = re.compile("[一-龠]")
                        if r.match(word):
                            self.__add_dic(word, array[self.category_num] * 3)
                        else:
                            self.__add_dic(word, array[self.category_num])

        file.close
        self.category_num += 1
        self.__output__(self.category_num, category, self.dic)
        return self.dic

    def __add_dic(self, word, count):
        if word in self.dic:
            self.dic[word] += count
        else:
            self.dic[word] = count

    def __read_from_exclude(self):
        file = open(self.exclude, 'r')
        for line in file:
            exclude_word = line.rstrip().split("\t")
            self.exclude_list.append(exclude_word)
        file.close
        return self.exclude_list

if __name__=='__main__':
    if sys.version_info > (3,0):
        if len(sys.argv) > 2:
            analyzer = Analyzer(sys.argv)
            analyzer.start()
        else:
            print("Invalid arguments")
    else:
        print("This program require python > 3.0")

